src/virtio/net.c

```diff
@@ -0,0 +1,406 @@
+    if (respond_to_client) {
+        if (!virtio_net_respond(dev)) {
+            return 0;
+        }
+    }
+
+    return 1;
+}
+
+static uint32_t copy_rx(struct virtq *virtq,
+                        uint16_t *curr_desc_head,
+                        uint32_t *desc_copied,
+                        const void *buf, uint32_t size)
+{
+    uint32_t copied = 0;
+    do {
+        uint32_t copying = MIN(size - copied, virtq->desc[*curr_desc_head].len - *desc_copied);
+
+        memcpy((void *)virtq->desc[*curr_desc_head].addr + *desc_copied, buf + copied, copying);
+
+        copied += copying;
+        *desc_copied += copying;
+
+        if (*desc_copied == virtq->desc[*curr_desc_head].len) {
+            if (!(virtq->desc[*curr_desc_head].flags & VIRTQ_DESC_F_NEXT)) {
+                break;
+            }
+            *curr_desc_head = virtq->desc[*curr_desc_head].next;
+            *desc_copied = 0;
+        }
+    } while (copied < size);
+
+    return copied;
+}
+
+static void handle_rx_buffer(struct virtio_device *dev,
+                             uint64_t buf_offset, uint32_t size,
+                             bool *respond_to_client)
+{
+    struct virtio_net_device *state = device_state(dev);
+
+    virtio_queue_handler_t *vq = &dev->vqs[VIRTIO_NET_RX_VIRTQ];
+    struct virtq *virtq = &vq->virtq;
+
+    uint16_t guest_idx = virtq->avail->idx;
+    uint16_t idx = vq->last_idx;
+
+    if (idx == guest_idx) {
+        /* vq is full or not initialised, drop the packet */
+        return;
+    }
+
+    /* Read the head of the descriptor chain */
+    uint16_t desc_head = virtq->avail->ring[idx % virtq->num];
+    uint16_t curr_desc_head = desc_head;
+
+    uint32_t copied = 0;
+    /* Amount of the current descriptor copied */
+    uint32_t desc_copied = 0;
+
+    struct virtio_net_hdr_mrg_rxbuf virtio_hdr = {0};
+    virtio_hdr.num_buffers = 1;
+
+    copied += copy_rx(virtq, &curr_desc_head, &desc_copied, &virtio_hdr, sizeof(struct virtio_net_hdr_mrg_rxbuf));
+    copied += copy_rx(virtq, &curr_desc_head, &desc_copied, state->rx_data + buf_offset, size);
+
+    /* Put it in the used ring */
+    virtq_enqueue_used(virtq, desc_head, copied);
+
+    /* Record that we've used this descriptor chain now */
+    vq->last_idx++;
+
+    *respond_to_client = true;
+}
+
+void virtio_net_handle_rx(struct virtio_net_device *state)
+{
+    struct virtio_device *dev = &state->virtio_device;
+
+    if (!driver_ok(dev)) {
+        return;
+    }
+    if (!dev->vqs[VIRTIO_NET_RX_VIRTQ].ready) {
+        /* vq is not initialised, drop the packet */
+        return;
+    }
+
+    net_buff_desc_t sddf_buffer;
+    bool reprocess = true;
+    bool respond_to_client = false;
+
+    while (reprocess) {
+        while (net_dequeue_active(&state->rx, &sddf_buffer) != -1) {
+            /* On failure, drop packet since we don't know how long until next interrupt */
+            handle_rx_buffer(dev, sddf_buffer.io_or_offset, sddf_buffer.len, &respond_to_client);
```

Comment:
What about having the loop condition check that `virtq->avail->idx != last_idx` and then removing this whole `respond_to_client` pointer.

Reply:
Discussed offline, won't do the refactor.